File: src/common/Configuration/YamlConfigFile.cs
using System;
using System.Collections.Generic;
using System.IO;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

public class YamlConfigFile : ConfigFile
{
    public YamlConfigFile(string filePath, ConfigFileScope scope) : base(filePath, scope)
    {
    }

    protected override Dictionary<string, object> ReadSettings(string fileName)
    {
        var exists = FileHelpers.FileExists(fileName);
        if (!exists) return new Dictionary<string, object>();

        try
        {
            var deserializer = new DeserializerBuilder()
                .WithNamingConvention(PascalCaseNamingConvention.Instance)
                .Build();

            var yamlContent = FileHelpers.ReadAllText(fileName);
            var result = deserializer.Deserialize<Dictionary<string, object>>(yamlContent) ?? new Dictionary<string, object>();
            
            return NormalizeNestedDictionaries(result);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error reading YAML configuration file: {ex.Message}");
            return new Dictionary<string, object>();
        }
    }

    protected override void WriteSettings(string fileName, Dictionary<string, object>? settings = null)
    {
        try
        {
            var serializer = new SerializerBuilder()
                .WithNamingConvention(PascalCaseNamingConvention.Instance)
                .Build();

            var yamlContent = serializer.Serialize(Settings);
            FileHelpers.WriteAllText(fileName, yamlContent);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error writing YAML configuration file: {ex.Message}");
        }
    }

    private Dictionary<string, object> NormalizeNestedDictionaries(Dictionary<string, object> data)
    {
        var result = new Dictionary<string, object>();

        foreach (var pair in data)
        {
            if (pair.Value is Dictionary<object, object> nestedDict)
            {
                // Convert Dictionary<object, object> to Dictionary<string, object>
                var convertedDict = new Dictionary<string, object>();
                foreach (var nestedPair in nestedDict)
                {
                    var key = nestedPair.Key.ToString() ?? string.Empty;
                    if (nestedPair.Value is Dictionary<object, object> deepNestedDict)
                    {
                        // Handle deeply nested dictionaries recursively
                        convertedDict[key] = NormalizeNestedDictionaries(
                            ConvertToStringDictionary(deepNestedDict));
                    }
                    else
                    {
                        convertedDict[key] = nestedPair.Value ?? string.Empty;
                    }
                }
                result[pair.Key] = convertedDict;
            }
            else
            {
                result[pair.Key] = pair.Value ?? string.Empty;
            }
        }

        return result;
    }

    private Dictionary<string, object> ConvertToStringDictionary(Dictionary<object, object> dict)
    {
        var result = new Dictionary<string, object>();
        foreach (var pair in dict)
        {
            var key = pair.Key.ToString() ?? string.Empty;
            result[key] = pair.Value ?? string.Empty;
        }
        return result;
    }
}

File: src/commonjs/src/Configuration/YamlConfigFile.ts
import * as fs from 'fs';
import * as yaml from 'js-yaml';
import { ConfigFile } from './ConfigFile';
import { ConfigFileScope } from './ConfigFileScope';

export class YamlConfigFile extends ConfigFile {
    constructor(filePath: string, scope: ConfigFileScope) {
        super(filePath, scope);
    }

    protected readSettings(fileName: string): Record<string, any> {
        if (!fs.existsSync(fileName)) {
            return {};
        }

        try {
            const yamlContent = fs.readFileSync(fileName, 'utf-8');
            const result = yaml.load(yamlContent) as Record<string, any> || {};
            return this.normalizeNestedDictionaries(result);
        } catch (error) {
            console.error(`Error reading YAML configuration file: ${(error as Error).message}`);
            return {};
        }
    }

    protected writeSettings(fileName: string, settings?: Record<string, any>): void {
        try {
            const yamlContent = yaml.dump(this.settings, {
                indent: 2,
                lineWidth: -1,
                noRefs: true
            });
            fs.writeFileSync(fileName, yamlContent);
        } catch (error) {
            console.error(`Error writing YAML configuration file: ${(error as Error).message}`);
        }
    }

    private normalizeNestedDictionaries(data: Record<string, any>): Record<string, any> {
        const result: Record<string, any> = {};

        for (const [key, value] of Object.entries(data)) {
            if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
                // Recursively normalize nested objects
                result[key] = this.normalizeNestedDictionaries(value);
            } else {
                result[key] = value ?? '';
            }
        }

        return result;
    }
}